server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -278,7 +280,18 @@ public long buildTookInMillis() {
 
     @Override
     protected void doExecute(Task task, SearchRequest searchRequest, ActionListener<SearchResponse> listener) {
-        executeRequest((SearchTask) task, searchRequest, listener, AsyncSearchActionProvider::new);
+        ActionListener<SearchResponse> loggingListener = listener.delegateFailureAndWrap((l, searchResponse) -> {
+            if (searchResponse.getShardFailures() != null && searchResponse.getShardFailures().length > 0) {
+                // Deduplicate failures by exception message and index
+                ShardOperationFailedException[] groupedFailures = ExceptionsHelper.groupBy(searchResponse.getShardFailures());
+                for (ShardOperationFailedException f : groupedFailures) {
+                    if (f.status().getStatus() >= 500 && ExceptionsHelper.isNodeOrShardUnavailableTypeException(f.getCause()) == false)
```

Comment:
I think we should try and get the cause of the ShardOperationFailedException to base the determination of the status node on. I think that is what we want to do in other places like in #101750. When I just introduced intentional NPEs in field mappers that should fail on query rewrite, the logging didn't trigger.

Reply:
One additional note: when I changed this so the status check uses the cause of the ShardOperationFailedException I noticed we still wouldn't see anything bubbling up from any lucene query creation on the shards (i.e everything that happens in SearchExecutionContext#toQuery) since we [catch everything there](https://github.com/elastic/elasticsearch/blob/61c7483fc930268cc1ec2b791abfc253be5d9171/server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java#L465) and wrap it into a QueryShardException which in turn is later [wrapped into a SearchShardFailure](https://github.com/elastic/elasticsearch/blob/f3a4813159869f910b5f6255c1e207f517f18811/server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java#L548) which is what we see here. When I deliberately introduced NPEs in field mapper code, this wasn't logged here even after getting the ShardOperationFailedException cause.

I don't think this is a problem really, but it shows we won't be able to see every potential problem even with this logging.